XDMA/linux-kernel/libxdma/libxdma.h

```diff
@@ -66,7 +66,7 @@
 #define XDMA_OFS_CONFIG		(0x3000UL)
 
 /* maximum number of desc per transfer request */
-#define XDMA_TRANSFER_MAX_DESC (2048)
+#define XDMA_TRANSFER_MAX_DESC (512)
```

Comment:
was this a necessary change for the default configuration? can you comment on why this is necessary?

Reply:
I thinks this change isn't nessary. It didn't solve my problem (see https://github.com/Xilinx/dma_ip_drivers/issues/54).

Reply:
The documentation from Xilinx (https://www.xilinx.com/support/documentation/ip_documentation/xdma/v4_1/pg195-pcie-dma.pdf) notes that Gen3x16 has a descriptor FIFO depth of 1024 and Gen3x8 (which is what we're using) is 512. I'm having trouble interpretting the language on pages 24 and 25, but it sounds like the engine might be smart about not overflowing this FIFO. What's certain is that the engine can't possibly fetch more than the FIFO allows at once, so I believe the value I've shown is safest for all hardware.

Reply:
There is a message of a race condition here:
https://forums.aws.amazon.com/thread.jspa?threadID=295643

ref to
 timeout_ms 

Maybe your review and dev knowledge may get some ideas...

Reply:
I would open a new issue for that. It doesn't seem related to this. You need details like which commit hash you're using and the sample program that caused the error. For what it's worth, it seems odd to me that the engine apparently times out after only one descriptor completed (see completed_desc_count). Also, FYI, error 512 is ERESTARTSYS and is explicitly returned by xdma_xfer_submit on timeout.

Reply:
> I would open a new issue for that. It doesn't seem related to this. You need details like which commit hash you're using and the sample program that caused the error. For what it's worth, it seems odd to me that the engine apparently times out after only one descriptor completed (see completed_desc_count). Also, FYI, error 512 is ERESTARTSYS and is explicitly returned by xdma_xfer_submit on timeout.

Hello,
I have opened a issue (https://github.com/Xilinx/dma_ip_drivers/issues/54) for I got the error 512 when I transfer 1MB C2H in stream mode.

Reply:
i tried the changes by bhathaway. I also experimented with various numbers down to 

XDMA_TRANSFER_MAX_DESC (16).

Job works for a few minutes and hangs eventually.  
Ubuntu 18.04.02

Reply:
>I would be genuinely curious if anyone has benchmarked the performance at different values for this constant.

Quite an old topic but without benchmarking, I can tell that testing using a loopback channel (3x16), I see the following:
* If I queue 1024 descriptors on the C2H channel, and then queue write descriptors, the transfers are not done. If queuing 1023 read descriptors, it works. I think it means that the choice of this number needs to be done depending on the application. If there is no way reads could lock writes, then no problem, if on the contrary, in some cases, reads cannot happen before a write is done (like loopback mode), this number should not exceed 1023 (on the 3x16 IP).

* To test this, fetch my adjacent descriptors fix (without that, loopback transfers for more than 1 page mostly don't work or crash the machine because of crossing page boundaries). Use the dma_streaming script and test with sizes up to 1023 pages (4190208 bytes). They work.

* Now test with more than 4190208, the transfer locks up.

* Reducing the max number of descriptors to 1023 makes the transfers work in all cases.

I cannot say for sure that the IP works the way I think it does but I'm just putting these here for others to possibly test/confirm or just to help them.